x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/slowlog/EsqlSlowLog.java

```diff
@@ -0,0 +1,95 @@
+
+    public EsqlSlowLog(ClusterSettings settings) {
+        this.clusterSettings = settings;
+    }
+
+    public void onQueryPhase(Result esqlResult, String query) {
+        if (esqlResult == null) {
+            return; // TODO review, it happens in some tests, not sure if it's a thing also in prod
+        }
+        long tookInNanos = esqlResult.executionInfo().overallTook().nanos();
+        var queryWarnThreshold = clusterSettings.get(ESQL_SLOWLOG_THRESHOLD_QUERY_WARN_SETTING).nanos();
+        if (queryWarnThreshold >= 0 && tookInNanos > queryWarnThreshold) {
+            queryLogger.warn(Message.of(esqlResult, query));
+            return;
+        }
+        var queryInfoThreshold = clusterSettings.get(ESQL_SLOWLOG_THRESHOLD_QUERY_INFO_SETTING).nanos();
+        if (queryInfoThreshold >= 0 && tookInNanos > queryInfoThreshold) {
+            queryLogger.info(Message.of(esqlResult, query));
+            return;
+        }
+        var queryDebugThreshold = clusterSettings.get(ESQL_SLOWLOG_THRESHOLD_QUERY_DEBUG_SETTING).nanos();
+        if (queryDebugThreshold >= 0 && tookInNanos > queryDebugThreshold) {
+            queryLogger.debug(Message.of(esqlResult, query));
+            return;
+        }
+        var queryTraceThreshold = clusterSettings.get(ESQL_SLOWLOG_THRESHOLD_QUERY_TRACE_SETTING).nanos();
+        if (queryTraceThreshold >= 0 && tookInNanos > queryTraceThreshold) {
+            queryLogger.trace(Message.of(esqlResult, query));
+            return;
+        }
+    }
+
+    static final class Message {
+
+        private static String escapeJson(String text) {
+            byte[] sourceEscaped = JsonStringEncoder.getInstance().quoteAsUTF8(text);
+            return new String(sourceEscaped, StandardCharsets.UTF_8);
+        }
+
+        public static ESLogMessage of(Result esqlResult, String query) {
+            Map<String, Object> jsonFields = prepareMap(esqlResult, query);
+
+            return new ESLogMessage().withFields(jsonFields);
+        }
+
+        private static Map<String, Object> prepareMap(Result esqlResult, String query) {
+            Map<String, Object> messageFields = new HashMap<>();
+            messageFields.put("elasticsearch.slowlog.message", esqlResult.executionInfo().clusterAliases());
+            messageFields.put("elasticsearch.slowlog.took", esqlResult.executionInfo().overallTook().nanos());
+            messageFields.put("elasticsearch.slowlog.took_millis", esqlResult.executionInfo().overallTook().millis());
+            messageFields.put("elasticsearch.slowlog.planning.took", esqlResult.executionInfo().planningTookTime().nanos());
+            messageFields.put("elasticsearch.slowlog.planning.took_millis", esqlResult.executionInfo().planningTookTime().millis());
+            messageFields.put("elasticsearch.slowlog.search_type", "ESQL");
+            String source = escapeJson(query);
+            messageFields.put("elasticsearch.slowlog.source", source);
```

Comment:
I think source is a bit ambiguous. Should we call it query here?

Reply:
Yeah, I don't like it very much, `query` would be much better.

Reply:
I thought about this, source for _search is the actual request body, while in our case it's only the query string.
I'll rename it as "query" for now.